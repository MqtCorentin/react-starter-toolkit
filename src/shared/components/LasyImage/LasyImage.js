import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import loaderSVG from 'shared/images/puff.svg';
import './LasyImage.scss';

export const loadImage = name => import(`shared/images/${name}`);

const Loading = () => <img className="af-lasyimage" src={loaderSVG} alt="loading..." />;

export const useLoadImage = ({ name, loadImageFn = loadImage, initState = null }) => {
  const [stateimage, setStateImage] = useState(initState);

  useEffect(() => {
    if (stateimage === null) {
      loadImageFn(name)
        .then(image => setStateImage(image.default))
        .catch(error => setStateImage(error));
    }
  }, [stateimage, loadImageFn, name]);

  return { stateimage, setStateImage };
};

const LasyImage = ({ name, alt, useLoadImageFn, ...rest }) => {
  const { stateimage } = useLoadImageFn({ name });
  return stateimage !== null ? <img {...rest} src={stateimage} alt={alt} /> : <Loading />;
};

export const menuEnhancedPropTypes = {
  useLoadImageFn: PropTypes.func,
};

export const menuEnhancedDefaultProps = {
  useLoadImageFn: useLoadImage,
};

LasyImage.propTypes = menuEnhancedPropTypes;
LasyImage.defaultProps = menuEnhancedDefaultProps;

export default LasyImage;
