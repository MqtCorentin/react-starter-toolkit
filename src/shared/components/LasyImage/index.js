import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import loaderSVG from 'shared/images/puff.svg';

export const loadImage = name => import(`shared/images/${name}`);

const Loading = () => (
  <img style={{ position: 'absolute', transform: 'translate(-50%,-50%)', left: '50%', top: '50%' }} src={loaderSVG} alt="loading..." />
);

export const useLoadImage = ({ name, loadImageFn = loadImage, initState = null }) => {
  const [stateimage, setStateImage] = useState(initState);

  useEffect(() => {
    if (stateimage === null) {
      loadImageFn(name)
        .then(image => setStateImage(image.default))
        .catch(error => setStateImage(error));
    }
  }, [stateimage, loadImageFn, name]);

  return { stateimage, setStateImage };
};

const LasyImage = ({ name, alt, useLoadImageFn, ...rest }) => {
  const { stateimage } = useLoadImageFn({ name });
  return stateimage !== null ? <img {...rest} src={stateimage} alt={alt} /> : <Loading />;
};

export const menuEnhancedPropTypes = {
  useLoadImageFn: PropTypes.func,
};

export const menuEnhancedDefaultProps = {
  useLoadImageFn: useLoadImage,
};

LasyImage.propTypes = menuEnhancedPropTypes;
LasyImage.defaultProps = menuEnhancedDefaultProps;

export default LasyImage;
