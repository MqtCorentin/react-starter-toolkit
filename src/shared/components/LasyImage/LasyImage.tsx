import { useState, useEffect } from 'react';
import loaderSVG from 'shared/images/puff.svg';
import './LasyImage.scss';

type TloadImage = string;
export const loadImage = (name: TloadImage) => import(`shared/images/${name}`);

const Loading = () => <img className="af-lasyimage" src={loaderSVG} alt="loading..." />;

type TuseLoadImage = {
  name: TloadImage;
  loadImageFn?: typeof loadImage;
  initState?: string | null;
};
export const useLoadImage = ({ name, loadImageFn = loadImage, initState = null }: TuseLoadImage) => {
  const [stateimage, setStateImage] = useState(initState);

  useEffect(() => {
    if (stateimage === null && name) {
      loadImageFn(name)
        .then(image => setStateImage(image.default))
        .catch(error => setStateImage(error));
    }
  }, [stateimage, loadImageFn, name]);

  return { stateimage, setStateImage };
};

type TLasyImage = {
  name: TloadImage;
  alt: string;
  useLoadImageFn?: typeof useLoadImage;
};

const LasyImage = ({ name, alt, useLoadImageFn = useLoadImage, ...rest }: TLasyImage) => {
  const { stateimage } = useLoadImageFn({ name });
  return stateimage !== null ? <img {...rest} src={stateimage} alt={alt} /> : <Loading />;
};

export default LasyImage;
