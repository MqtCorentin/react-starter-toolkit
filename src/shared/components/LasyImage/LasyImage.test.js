import React from 'react';
import { render, act } from '@testing-library/react';
import LasyImage from './index';

const defaultProps = { alt: 'alt', name: 'image.jpg' };

describe('<LasyImage/>', () => {
  it('Should render LasyImage', async () => {
    await act(async () => {
      const { asFragment } = render(<LasyImage {...defaultProps} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
