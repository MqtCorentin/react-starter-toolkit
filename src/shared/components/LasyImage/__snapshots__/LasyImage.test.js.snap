// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LasyImage/> Should render LasyImage 1`] = `
<DocumentFragment>
  <img
    alt="loading..."
    class="af-lasyimage"
    src="puff.svg"
  />
</DocumentFragment>
`;
